src/components/dialog/dialogHelpers.ts

```diff
@@ -0,0 +1,97 @@
+import type { ButtonKind } from '../button/props'
+import DialogHelper from './dialogsContainer.svelte'
+import { writable } from 'svelte/store'
+
+let helper: DialogHelper | undefined
+export const ensureDialogHelper = () => {
+  if (helper) return
+
+  const mountPoint = document.createElement('div')
+  mountPoint.id = 'leo-dialog-helper'
+
+  document.body.appendChild(mountPoint)
+
+  helper = new DialogHelper({
+    target: mountPoint
+  })
+}
+
+interface Action {
+  result?: any
+  text: string
+  kind?: ButtonKind
+}
+
+interface AlertOptions {
+  title: string
+  body?: string
+  actions: Action[]
+}
+
+interface AlertInfo extends AlertOptions {
+  resolve: (result?: any) => void
+}
+
+export const dialogs = writable<AlertInfo[]>([])
+
+export const dialog = <T>(options: AlertOptions) => {
+  ensureDialogHelper()
+
+  let resolve: (result?: T) => void
+  const promise = new Promise((a) => (resolve = a))
+
+  const info = {
+    title: options.title,
+    body: options.body,
+    actions: options.actions,
+    resolve: (result?: T) => {
+      dialogs.update((d) => d.filter((i) => i !== info))
+      resolve(result)
+    }
+  }
+
+  dialogs.update((d) => [...d, info])
```

Comment:
Technically this part could be done inside the Promise callback, could it not, avoiding both the `resolve` assignment and the possible but extremely unlikely possibility that resolve is called before it's assigned. Although I'm not sure when the Promise callback is called so it _might_ actually be impossible that happens...

Reply:
I think the earliest a promise can resolve is in the next tick, so it shouldn't ever happen. That said, yours reads nicer :laughing: